### PR TITLE
Add bounding box converter

### PR DESCRIPTION
**Pull request by qubvel**
# What does this PR do?

Intorduces `output_bbox_format` parameter for `postprocess_for_object_detection`.
Adds the `convert_boxes` function to convert boxes from one format to another, e.g. from YOLO to Pascal VOC.

```python
boxes = convert_boxes(
    boxes, input_format="relative_xcycwh", output_format="absolute_xyxy", image_size=target_sizes
)
```

 - works with `np.array`, `torch.tensor`, `list`, `tuple` (could be easily extended to `jax` and `tf`)
 - works with single box shape=(4,) , image boxes shape=(N, 4), multiple images boxes List[(N, 4)]
 - supports boxes with extra metadata, e.g. `box=[x_min, y_min, x_max, y_max, color, class, ...]`
 - supports format aliases. e.g. `yolo`, `coco`
 
Supported input/output bounding box formats:
    - `absolute_xyxy` (aliases: `pascal_voc`, `xyxy`): [x_min, y_min, x_max, y_max]
    - `absolute_xywh` (aliases: `coco`, `xywh`): [x_min, y_min, width, height]
    - `absolute_xcycwh`: [center_x, center_y, width, height]
    - `relative_xyxy` (aliases: `albumentations`): [x_min, y_min, x_max, y_max] normalized to [0, 1] by image size
    - `relative_xywh`: [x_min, y_min, width, height] normalized to [0, 1] by image size
    - `relative_xcycwh` (aliases: `yolo`, `xcycwh`): [center_x, center_y, width, height] normalized to [0, 1] by image size

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@amyeroberts could you please look at this draft, looks a bit overсomplicated, probably we should drop some functionality in favor to simplification.

### CODE REVIEW

File: src/transformers/models/grounding_dino/image_processing_grounding_dino.py [hunk 1]
```diff
@@ -193,17 +192,16 @@ def safe_squeeze(arr: np.ndarray, axis: Optional[int] = None) -> np.ndarray:
...
-            boxes = value
-            boxes = corners_to_center_format(boxes)
-            boxes /= np.asarray([image_width, image_height, image_width, image_height], dtype=np.float32)
-            norm_annotation[key] = boxes
-        else:
-            norm_annotation[key] = value
-    return norm_annotation
+    """Convert annotation boxes from absolute xyxy (pascal voc) to relative xcycwh (yolo) format."""
+    if "boxes" in annotation:
+        annotation = annotation.copy()  # shallow copy
```
Comment (amyeroberts) [inline src/transformers/models/grounding_dino/image_processing_grounding_dino.py:197]: What's the purpose of the copy here? Only `annotation["boxes"]` is being passed to `convert_boxes` and is being set directly to `annotation["boxes"]` on the call

File: src/transformers/image_box_utils.py [hunk 2]
```diff
@@ -0,0 +1,523 @@
...
+    import torch
+
+
+# TODO: align with transformers logging
+logger = logging.getLogger(__name__)
+
+ArrayType = Union["torch.Tensor", np.ndarray]
+
+
+SUPPORTED_BOX_FORMATS = [
```
Comment (amyeroberts) [inline src/transformers/image_box_utils.py:23]: I'd have this as an enum

File: src/transformers/image_box_utils.py [hunk 3]
```diff
@@ -0,0 +1,523 @@
...
+        - A single box (torch.Tensor | np.ndarray | List[float, ...]): Box of shape (B,), where B is 4 or more elements,
+        and the first 4 elements are the coordinates of the box.
+        - Boxes from one image (torch.Tensor | np.ndarray | List[List[float, ...]]): A set of boxes (N, B) where N is
+        the number of boxes.
+        - Boxes from multiple images (torch.Tensor | np.ndarray | List[torch.Tensor | np.ndarray] | List[List[List[float, ...]]]):
+        A set of images with their boxes (I, N, B), where I is number of images. Can be represented as a single 3D array/tensor
+        or most likely as a list of 2D arrays/tensors.
+
+    Supported input/output bounding box formats:
+        - `absolute_xyxy` (aliases: `pascal_voc`, `xyxy`): [x_min, y_min, x_max, y_max]
```
Comment (amyeroberts) [inline src/transformers/image_box_utils.py:491]: I'd rather we didn't try to handle different aliases and just keep it simple to e.g. `xyxy`

File: src/transformers/image_box_utils.py [hunk 4]
```diff
@@ -0,0 +1,523 @@
...
+    else:
+        raise ValueError(f"Unsupported type of boxes: {type(boxes)}")
+
+
+def convert_boxes(
+    boxes: Union[ArrayType, List, Tuple],
+    input_format: str,
+    output_format: str,
+    image_size: Optional[Union[ArrayType, List, Tuple]] = None,
+    check: Optional[str] = "warn",
```
Comment (amyeroberts) [inline src/transformers/image_box_utils.py:476]: I can see `"raise"` never being used. I like the logic, but think we might be engineering something which we'll never need

File: src/transformers/image_box_utils.py [hunk 5]
```diff
@@ -0,0 +1,523 @@
...
+        raise ValueError(f"Expected boxes to have at least 4 elements, got {len(boxes)} elements.")
+    if is_array_type(boxes) and _shape(boxes)[-1] < 4:
+        raise ValueError(f"Expected boxes to have at least 4 elements, got {_shape(boxes)[-1]} elements.")
+
+    # Base of recursion.
+    if depth == 2 and is_array_type(boxes):
+        return _convert_boxes_arrays_2d(boxes, input_format, output_format, image_size, check)
+
+    # Recursive approach.
+    elif depth == 1 and isinstance(boxes, (list, tuple) or is_array_type(boxes)):
```
Comment (amyeroberts) [inline src/transformers/image_box_utils.py:442]: Regarding the ` isinstance(boxes, (list, tuple) or is_array_type(boxes))` check - what are the other types `boxes` could be here?

**Review by amyeroberts — COMMENTED** [review]
Thanks for working on this! 

I haven't done a full review e.g. checking boring things like docstring formats, just looking at the overall structure. 

This is really nicely structured and handled. I don't have any major comments on the overall logic - it's incredibly clean. Thanks for taking the time to write out comprehnsive tests too. Only major comment is to not have alias for the different formats, in particular for datasets like coco, it adds unnecessary complexity; I'm less sure about e.g. `xyxy` -> `absolute_xyxy`.

**Comment by amyeroberts** [comment]
Adding the WIP label so the stalebot doesn't bother
